qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,796 @@
+            # only single curve
+            return self.__fit_funcs__[0](xvals, *params)
+
+        for curve_properties in self.__series__:
+            if curve_properties.name == curve_name:
+
+                # remap parameters
+                series_params = curve_properties.param_names
+                mapped_params = []
+                for series_param in series_params:
+                    try:
+                        param_idx = self.__param_names__.index(series_param)
+                    except ValueError as ex:
+                        raise QiskitError(
+                            f"Local function parameter {series_param} is not defined in "
+                            f"this class. {series_param} not in {self.__param_names__}."
+                        ) from ex
+                    mapped_params.append(params[param_idx])
+
+                # find fit function
+                f_index = curve_properties.fit_func_index
+                try:
+                    return self.__fit_funcs__[f_index](xvals, *mapped_params)
+                except IndexError as ex:
+                    raise QiskitError(f"Fit function of index {f_index} is not defined.") from ex
+
+        raise QiskitError(f"A curve {curve_name} is not defined in this class.")
+
+    def _run_analysis(self, experiment_data: ExperimentData, **options):
+        """Run analysis on circuit data.
+
+        Args:
+            experiment_data: The experiment data to analyze.
+            options: kwarg options for analysis function. This may contain `data_processor` key
+                as a special argument so that user can override default data processor.
+
+        Returns:
+            tuple: A pair ``(analysis_results, figures)`` where
+                   ``analysis_results`` may be a single or list of
+                   AnalysisResult objects, and ``figures`` may be
+                   None, a single figure, or a list of figures.
+        """
+        analysis_result = AnalysisResult()
+
+        # Setup data processor
+        if "data_processor" in options:
+            # Use user provided processor
+            data_processor = options.pop("data_processor")
+        else:
+            # Use default processor
+            if isinstance(self.__default_data_processor__, DataProcessor):
+                # Calibrate default data processor instance
+                try:
+                    data_processor = self._calibrate_data_processor(
+                        data_processor=self.__default_data_processor__,
+                        experiment_data=experiment_data,
+                    )
+                except DataProcessorError as ex:
+                    analysis_result["error_message"] = str(ex)
+                    analysis_result["success"] = False
+                    return analysis_result, list()
+            else:
+                # Callback function
+                data_processor = self.__default_data_processor__.__func__
```

Comment:
This seems to assume that the user provided data processor is always calibrated. This may not always be the case. What if I want to train it on the data of this particular experiment? I think we can decouple the logic. How about:
```python
if "data_processor" in options:
    data_processor = options.pop("data_processor")
else:
    data_processor = self.__default_data_processor__
    
if isinstance(data_processor, DataProcessor):
    # try doing the training
```
As a side note: why do you need `self.__default_data_processor__.__func__` if data processors are callable?

Reply:
Yes, but this can override user-calibrated processor setting. Can the processor have option like `frozen`?

> why do you need self.__default_data_processor__.__func__ if data processors are callable?

Without this, the attached callable is recognized as an instance method. i.e. unexpected `self` is passed as a first argument.

Reply:
I'm thinking of adding a `trained` property to the data processor as well as a `train` method. In DataProcessor we could have something along the lines
```python
@property
def trained(self) -> bool:
    return self._trained

def train(self, data: List[Dict[str, Any]]):
    if self.trained:
        return
    else:
        # train each node one after the other
        
    self._trained = True
```
Each node would also need this. While not yet finished you can see the start of this logic in https://github.com/Qiskit/qiskit-experiments/pull/48 in class `SVDAvg`. Once we have this we could then do in curve fit
```python
if isinstance(data_processor, DataProcessor):
    if not data_processor.trained:
        data_processor.train(data)
```

Reply:
This looks reasonable. I'll update the logic.